src/priceClient/priceClient.ts

```diff
@@ -113,44 +111,49 @@ export class PriceClient implements PriceFeedAdapter {
     return this.prices[currency];
   }
 
-  private async requestPrices(addresses: string[], currency: string): Promise<PriceCache> {
-    let prices: TokenPrice[] = [];
+  private async updatePrices(currency: string): Promise<void> {
+    const priceCache = this.getPriceCache(currency);
+    let addresses = Object.keys(priceCache);
 
     for (const priceFeed of this.priceFeeds) {
       this.logger.debug({
-        at: "PriceClient#getPriceByAddress",
+        at: "PriceClient#updatePrices",
         message: `Looking up prices via ${priceFeed.name}.`,
+        tokens: addresses,
       });
       try {
-        prices = await priceFeed.getPricesByAddress(addresses, currency);
-        if (prices.length === 0) throw Error("Zero-length response received");
-        // @todo: Handle partial price retrievals.
-        else if (prices.length === addresses.length) break;
+        const prices = await priceFeed.getPricesByAddress(addresses, currency);
+        addresses = await this.updateCache(priceCache, prices, addresses);
+        if (addresses.length === 0) break; // All done
       } catch (err) {
         this.logger.debug({
-          at: "PriceClient#requestPrices",
+          at: "PriceClient#updatePrices",
           message: `Price lookup against ${priceFeed.name} failed (${err}).`,
           tokens: addresses,
         });
         // Failover to the next price feed...
       }
     }
 
-    if (prices.length === 0) {
+    if (addresses.length !== 0) {
+      this.logger.debug({
+        at: "PriceClient#updatePrices",
+        message: "Unable to resolve some token prices.",
+        priceFeeds: this.listPriceFeeds(),
+        tokens: addresses,
+      });
```

Comment:
This might seem like overkill/duplication given the exception being thrown below. I wanted to avoid making the exception error message too long and convoluted. Throwing an object might make more sense, but it seems like that should be a class-wide pattern, rather than an isolated case.